### PR TITLE
Fixing m4t.

### PR DESCRIPTION
**Pull request by Narsil**
# What does this PR do?

Fixed m4t which had a disparity in configuration for SpeechToText.

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/models/seamless_m4t/modeling_seamless_m4t.py [hunk 1]
```diff
@@ -3051,8 +3051,9 @@ class SeamlessM4TForSpeechToText(SeamlessM4TPreTrainedModel):
     def __init__(self, config: SeamlessM4TConfig):
         super().__init__(config)
 
+        self.shared = nn.Embedding(config.vocab_size, config.hidden_size, config.pad_token_id)
         self.speech_encoder = SeamlessM4TSpeechEncoder(config)
-        self.text_decoder = SeamlessM4TDecoder(config)
+        self.text_decoder = SeamlessM4TDecoder(config, self.shared)
```
Comment (ylacombe) [inline src/transformers/models/seamless_m4t/modeling_seamless_m4t.py:3056]: If I understand it correctly, this should also appears on the `SeamlessM4TForSpeechToSpeech` class right ?

**Review by ylacombe — APPROVED** [review]
LGTM! Thanks for fixing the issue here and enriching the tests!

I left a comment because I believe we should do the same with `SeamlessM4TForSpeechToSpeech`, WDYT ?

File: src/transformers/models/seamless_m4t/modeling_seamless_m4t.py (hunk 1 above)
Comment (Narsil) [reply src/transformers/models/seamless_m4t/modeling_seamless_m4t.py:3056]: I'm not familiar with the models, but Iḿ guessing the text vocabulary is not needed there.

Comment (ylacombe) [reply src/transformers/models/seamless_m4t/modeling_seamless_m4t.py:3056]: The speech encoder doesn't have an embedding layer so need for a shared embeddings with it. The text decoder still needs it though

Comment (Narsil) [reply src/transformers/models/seamless_m4t/modeling_seamless_m4t.py:3056]: You were correct I didn't read properly that model I Supposed the SpeechDecoder decoded directly to voice, but it's actually being vocoded from the text, you are correct. However my tests do pass locally... Very odd.

Comment (Narsil) [reply src/transformers/models/seamless_m4t/modeling_seamless_m4t.py:3056]: OMG the test passes on CUDA but not CPU..... Witchcraft

**Review by LysandreJik — APPROVED** [review]
Thank you @Narsil ! Will merge when green

**Comment by LysandreJik** [comment]
Thanks @Narsil!

**Pull request merged**